datastore/src/main/scala/org/allenai/datastore/Datastore.scala

```diff
@@ -38,6 +40,27 @@ import scala.collection.JavaConverters._
   * @param s3   properly authenticated S3 client.
   */
 class Datastore(val name: String, val s3: AmazonS3Client) extends Logging {
+
+  private val random = new Random
+  private def withRetries[T](activity: String, retries: Int = 10)(f: => T): T =
+    if (retries <= 0) {
+      f
+    } else {
+      val sleepTime = random.nextInt(10000) + 10000 // sleep between 10 and 20 seconds
```

Comment:
If you're feeling API-ish, you could make the sleep time be an optional constructor parameter along with the starting retry count.